### PR TITLE
CMake add_phare_test to allow runtime test execution method override,…

### PR DESCRIPTION
**Pull request by Dekken**
i.e. with mpirun

mpirun tests being added to Teamcity to follow

### CODE REVIEW

**Comment by Dekken** [comment]
MPI tests are configured but disabled.

they fail as expected

```
[13:03:11][Step 5/5] The following tests FAILED:
[13:03:11][Step 5/5] 	  3 - test-particles-data-refine (Failed)
[13:03:11][Step 5/5] Errors while running CTest
[13:03:11][Step 5/5] 	  4 - test-particles-data-schedule (Failed)
[13:03:11][Step 5/5] 	  5 - test-field-data-coarsen (Failed)
[13:03:11][Step 5/5] 	 14 - test-messenger (Failed)
[13:03:11][Step 5/5] 	 16 - test-multiphysics-integrator (Failed)
```

https://hephaistos.lpp.polytechnique.fr/teamcity/viewLog.html?buildId=49534&buildTypeId=Phare_Phare_BuildGithubPrClang&tab=buildLog

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,6 +14,7 @@ endif()
 set(PHARE_PROJECT_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 
 option(test "Build test with google test" ON)
+option(testMPI "Build test with google test" OFF) # 2 processes on one host
 option(coverage "Generate coverage" ON)
```
Comment (nicolasaunai) [inline CMakeLists.txt]: can we also put coverage OFF by default? We'd have to put it on manually on TC

File: CMakeLists.txt [hunk 2]
```diff
@@ -34,6 +35,21 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS 1)
 if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
   set (CMAKE_CXX_FLAGS -fsized-deallocation)
 endif()
+
+if(coverage AND NOT MSVC)
+  set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pg -DHAVE_EXECINFO_H -g3")
+endif()
+if(testMPI)
+  function(add_phare_test binary directory)
+    add_test(NAME ${binary} COMMAND mpirun -n 2 ./${binary} WORKING_DIRECTORY ${directory})
```
Comment (nicolasaunai) [inline CMakeLists.txt]: do you think the number of processes could also be a parameter of the function? (would default=2 possible in that case?)

File: tests/amr/data/field/copy_pack/CMakeLists.txt [hunk 3]
```diff
@@ -26,5 +26,3 @@ target_link_libraries(${PROJECT_NAME} PUBLIC
 
 add_subdirectory(copy)
 add_subdirectory(copy_overlap)
-add_subdirectory(stream_pack)
```
Comment (nicolasaunai) [inline tests/amr/data/field/copy_pack/CMakeLists.txt]: the test now works on my branch `fix`

File: tests/amr/data/field/copy_pack/stream_pack/CMakeLists.txt [hunk 4]
```diff
@@ -10,8 +10,8 @@ set(SOURCES_INC
 
 set(SOURCES_CPP
   test_main.cpp
-#  test_stream_pack_centered_ex.cpp disabled cause it's b0rked
-#  test_stream_pack_centered_ey.cpp disabled cause it's b0rked
+  test_stream_pack_centered_ex.cpp disabled cause it's b0rked
```
Comment (nicolasaunai) [inline tests/amr/data/field/copy_pack/stream_pack/CMakeLists.txt]: so we can put it back once you account for the diffs in my branch `fix`

File: tests/amr/data/particles/CMakeLists.txt [hunk 5]
```diff
@@ -1,5 +1,4 @@
 add_subdirectory(copy)
 add_subdirectory(copy_overlap)
-add_subdirectory(stream_pack)
```
Comment (nicolasaunai) [inline tests/amr/data/particles/CMakeLists.txt]: back on

File: tests/amr/messengers/subcov.sh [hunk 6]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
```
Comment (nicolasaunai) [inline tests/amr/messengers/subcov.sh]: what is this file for? why is it needed for messenger test

File: tests/amr/messengers/test_basichierarchy.h [hunk 7]
```diff
@@ -67,15 +70,21 @@ class BasicHierarchy
               dimension_, "cartesian", inputDatabase_->getDatabase("CartesianGridGeometry"))}
         , hierarchy_{std::make_shared<SAMRAI::hier::PatchHierarchy>("PatchHierarchy", gridGeometry_,
                                                                     patchHierarchyDatabase_)}
-        , loadBalancer_{std::make_shared<SAMRAI::mesh::ChopAndPackLoadBalancer>(
-              dimension_, "ChopAndPackLoadBalancer",
-              inputDatabase_->getDatabase("ChopAndPackLoadBalancer"))}
+        // , loadBalancer_{std::make_shared<SAMRAI::mesh::ChopAndPackLoadBalancer>(
```
Comment (nicolasaunai) [inline tests/amr/messengers/test_basichierarchy.h]: :(

File: tests/amr/messengers/test_messengers.cpp [hunk 8]
```diff
@@ -46,9 +46,11 @@
 #include "gmock/gmock.h"
 #include "gtest/gtest.h"
 
+#include <chrono>
```
Comment (nicolasaunai) [inline tests/amr/messengers/test_messengers.cpp]: that and `thread` could probably be deleted

File: tests/coverage.sh [hunk 9]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
Comment (nicolasaunai) [inline tests/coverage.sh]: bash :(

File: tests/amr/data/field/copy_pack/stream_pack/CMakeLists.txt (hunk 4 above)
Comment (Dekken) [reply tests/amr/data/field/copy_pack/stream_pack/CMakeLists.txt]: are you sure ?

I have failures after merging

**Comment by Dekken** [comment]
obsoleted by https://github.com/PHAREHUB/PHARE/pull/56

**Pull request closed without merging**